compositor_render/src/render_loop.rs

```diff
@@ -0,0 +1,58 @@
+use std::{collections::HashMap, sync::Arc};
+
+use compositor_common::{scene::InputId, Frame};
+use log::error;
+
+use crate::renderer::{
+    scene::{Node, Scene},
+    texture::Texture,
+    RenderCtx,
+};
+
+pub(super) fn populate_inputs(
+    ctx: &RenderCtx,
+    scene: &Scene,
+    frames: &mut HashMap<InputId, Arc<Frame>>,
+) {
+    for (input_id, (_, yuv_textures)) in &scene.inputs {
+        Texture::upload_frame_to_textures(
+            ctx.wgpu_ctx,
+            yuv_textures,
+            frames.remove(input_id).unwrap(),
+        );
+        error!("TODO: convert yuv to rgba")
+    }
+}
+
+pub(super) fn run_transforms(ctx: &RenderCtx, scene: &Scene) {
+    for node in graph_iter(scene) {
+        let sources: Vec<_> = node
+            .inputs
+            .iter()
+            .map(|node| (&node.node_id, &node.output))
+            .collect();
+        node.transform.render(ctx, &sources, &node.output)
+    }
+}
+
+/// Returns iterator that guarantees that nodes will be visited in
+/// the order that ensures that inputs of the current node were
+/// already visited in previous iterations
+fn graph_iter(scene: &Scene) -> impl Iterator<Item = Arc<Node>> {
+    let mut queue = vec![];
+    for (output_node, _output_texture) in scene.outputs.values() {
+        enqueue_nodes(output_node, &mut queue);
+    }
+    queue.into_iter()
+}
+
+fn enqueue_nodes(parent_node: &Arc<Node>, queue: &mut Vec<Arc<Node>>) {
+    let already_in_queue = queue.iter().any(|i| Arc::ptr_eq(parent_node, i));
+    if already_in_queue {
+        return;
+    }
+    for child in parent_node.inputs.iter() {
+        enqueue_nodes(child, queue);
+    }
+    queue.push(parent_node.clone());
+}
```

Comment:
> * I don't think performance here matters

It doesn't matter on the level "hashmap lookup vs small list traverse", but I think it matters on the level "if our graph will have 30 nodes, will this blow up?" :P

> * Even if performance would matter, I suspect this solution would still be faster for the values n we care about.

Imagine the Membrane team joining the call on the Videoroom app using Compositor to combine streams. Let's say that we have 20 video camera inputs, transform each of them with 4 different transforms (like cropping, corners rounding, filter, etc.), and combine them into the grid. Now our scene graph has 101 nodes. At this scale, the difference between `O(n)`, `O(n^2)`, and `O(n^3)` isn't neglectable and this isn't the most complex scenario we can imagine.

Reply:
> At this scale, the difference between O(n), O(n^2), and O(n^3) isn't neglectable

At this scale (100 nodes) it's not obvious to me that Set would be faster than a n^3 solution. It most likely is but it's not obvious. Even if it's slower it would be measurable, but still a negligible difference.

Also, I suspect using recursive implementation would become a problem earlier. I don't expect any reasonable impact before 1000 nodes and series performance issues around 10k nodes